central/administration/events/datastore/internal/writer/writer_impl_test.go

```diff
@@ -0,0 +1,90 @@
+package writer
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pkg/errors"
+	storeMocks "github.com/stackrox/rox/central/administration/events/datastore/internal/store/mocks"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/administration/events"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/mock/gomock"
+)
+
+var errFake = errors.New("fake error")
+
+func TestEventsWriter(t *testing.T) {
+	t.Parallel()
+	suite.Run(t, new(writerTestSuite))
+}
+
+type writerTestSuite struct {
+	suite.Suite
+	mockCtrl *gomock.Controller
+
+	readCtx  context.Context
+	writeCtx context.Context
+	store    *storeMocks.MockStore
+	writer   Writer
+}
+
+func (s *writerTestSuite) SetupTest() {
+	s.mockCtrl = gomock.NewController(s.T())
+
+	s.readCtx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration),
+		),
+	)
+	s.writeCtx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration),
+		),
+	)
+	s.store = storeMocks.NewMockStore(s.mockCtrl)
+	s.writer = New(s.store)
+}
+
+func (s *writerTestSuite) TestWriteEvent_Error() {
+	event := &events.AdministrationEvent{
+		Level:   storage.AdministrationEventLevel_ADMINISTRATION_EVENT_LEVEL_ERROR,
+		Message: "message",
+		Type:    storage.AdministrationEventType_ADMINISTRATION_EVENT_TYPE_GENERIC,
+		Hint:    "hint",
+		Domain:  "domain",
+	}
+	id := "73072ecb-2222-5922-8948-5944338861c8"
+
+	s.store.EXPECT().Get(s.writeCtx, id).Return(nil, false, errFake)
+	err := s.writer.Upsert(s.writeCtx, event)
+	s.ErrorIs(err, errFake)
+}
+
+func (s *writerTestSuite) TestWriteEvent_NilEvent_Error() {
+	err := s.writer.Upsert(s.writeCtx, nil)
+	s.ErrorIs(err, errox.InvalidArgs)
+}
```

Comment:
I think it would be great to expand this test into a few test cases (or separate tests):
* `nil` event -> reject
* non-nil event but with all fields empty -> reject
* all fields that participate in UUID generation empty -> reject
* all fields that participate in UUID generation same -> merging should happen
* at least one field that participates in UUID generation is different -> no merging

Basically tests that cover expected behaviour of the writer.

Reply:
Hmm, I added a validation function to test most of the things you mentioned, some don't make sense to me / are already covered though.

>all fields that participate in UUID generation same -> merging should happen

This is covered within the postgres integration tests, `TestUpsertEvent_MultipleOccurrencesFlushOnce`. We validate that the occurrence is increased.

>at least one field that participates in UUID generation is different -> no merging
This is covered by the integration tests like the count test, where multiple different events are added and no merging happens.

>invalid event (probably should be a separate func close to event type declaration) -> reject (e.g., created_at <= last_occurred_at)

The timestamp is created within `AdministrationEvent.ToStorageEvent()` and set to `time.Now`. The timestamp cannot be modified from a caller, so do not think it is worth explicitly testing / calling out here.